### PR TITLE
Search job templates by name or description

### PR DESCRIPTION
**Pull request by kalikiana**
- Search job templates
- Render one or more templates and descriptions per job group

See: [poo#34486](https://progress.opensuse.org/issues/34486)

### CODE REVIEW

**Review by Martchus — COMMENTED** [review]
1. The "The search currently finds…" info hasn't been adjusted yet.
2. Now that we have multiple types of search results: It would be nice if the type of a search result would be displayed somewhere (e.g. in form of an icon with e.g. "Job template" as tooltip.)
3. It would also be nice to provide a link to the result (e.g. job template page or source code view).
4. As we add more and more things to search for, the `query` function should be splitted into multiple functions.

I'd say only 1. should be fixed within this PR.

**Comment by kalikiana** [comment]
> 1. The "The search currently finds…" info hasn't been adjusted yet.
> 2. Now that we have multiple types of search results: It would be nice if the type of a search result would be displayed somewhere (e.g. in form of an icon with e.g. "Job template" as tooltip.)
> 3. It would also be nice to provide a link to the result (e.g. job template page or source code view).
> 4. As we add more and more things to search for, the `query` function should be splitted into multiple functions.
> 
> I'd say only 1. needs to be fixed within this PR.

Ack. I updated the description. The other items will be separate (as also discussed in the call).

**Review by Martchus — APPROVED** [review]

**Review by Amrysliu — APPROVED** [review]
LGTM

File: lib/OpenQA/WebAPI/Controller/API/V1/Search.pm [hunk 1]
```diff
@@ -117,6 +117,23 @@ sub query {
...
         last if $cap < 1;
+
+        # Job templates
+        my $last_group = undef;
+        my $like       = {like => "%${keywords}%"};
+        my $templates  = $self->schema->resultset('JobTemplates')
+          ->search({-or => {name => $like, description => $like}}, {limit => $cap});
+        while (my $template = $templates->next) {
+            my $contents = $template->name . "\n" . $template->description;
+            if ($template->group->id == $last_group) {
```
Comment (Amrysliu) [inline lib/OpenQA/WebAPI/Controller/API/V1/Search.pm:128]: @kalikiana  I did not realize this $last_group would be `undef` until I did a test.
There is a message when I use the search function: `Use of uninitialized value $last_group in numeric eq (==) at /home/liuxiaojing/code/repos/openQA/script/../lib/OpenQA/WebAPI/Controller/API/V1/Search.pm line 128.`

**Comment by Amrysliu** [comment]
@kalikiana 
IMHO The search job templates should not be in the loop `for my $distri ($distris->list({dir => 1})->map('realpath')->uniq()->each)`.  Because the job templates are not related to the test directories.  And the result is the job template name is shown four times in my test (there are four test directories in my environment).
Sorry for that,  I did not find this while reviewing the code.

File: lib/OpenQA/WebAPI/Controller/API/V1/Search.pm (hunk 1 above)
Comment (Martchus) [reply lib/OpenQA/WebAPI/Controller/API/V1/Search.pm:128]: @Amrysliu is right. If one only looks at the diff it isn't very apparent that this is actually within the loop. This function really gets a little bit big.

Comment (kalikiana) [reply lib/OpenQA/WebAPI/Controller/API/V1/Search.pm:128]: I proposed #3335 as a follow-up, refactoring the result types into separate functions and addressing the above concerns in tandem.

**Pull request merged**